geomstats/geometry/poincare_ball.py

```diff
@@ -0,0 +1,313 @@
+                                    base_point.shape[-1], -1)
+
+        log = (1 - norm_base_point**2) * gs.arctanh(norm_add)\
+            * (add_base_point / norm_add)
+
+        mask_0 = gs.isclose(norm_add, 0.)
+        log[mask_0] = 0
+
+        return log
+
+    def mobius_add(self, point_a, point_b):
+        r"""Compute the Mobius addition of two points.
+
+        Mobius addition operation that is a necessary operation
+        to compute the log and exp using the 'ball' representation.
+
+        .. math::
+
+            a\oplus b=\frac{(1+2\langle a,b\rangle + ||b||^2)a+
+            (1-||a||^2)b}{1+2\langle a,b\rangle + ||a||^2||b||^2}
+
+        Parameters
+        ----------
+        point_a : array-like, shape=[n_samples, dimension]
+            Point in hyperbolic space.
+        point_b : array-like, shape=[n_samples, dimension]
+            Point in hyperbolic space.
+
+        Returns
+        -------
+        mobius_add : array-like, shape=[n_samples, 1]
+            Result of the Mobius addition.
+        """
+        norm_point_a = gs.sum(point_a ** 2, axis=-1,
+                              keepdims=True)
+
+        norm_point_a = gs.repeat(norm_point_a, point_a.shape[-1], -1)
+
+        norm_point_b = gs.sum(point_b ** 2, axis=-1,
+                              keepdims=True)
+        norm_point_b = gs.repeat(norm_point_b, point_a.shape[-1], -1)
+
+        sum_prod_a_b = gs.sum(point_a * point_b,
+                              axis=-1, keepdims=True)
+
+        sum_prod_a_b = gs.repeat(sum_prod_a_b, point_a.shape[-1], -1)
+
+        add_nominator = ((1 + 2 * sum_prod_a_b + norm_point_b) * point_a +
+                         (1 - norm_point_a) * point_b)
+
+        add_denominator = (1 + 2 * sum_prod_a_b + norm_point_a * norm_point_b)
+
+        mobius_add = add_nominator / add_denominator
+
+        return mobius_add
+
+    def dist(self, point_a, point_b):
+        """Compute the geodesic distance between two points.
+
+        Parameters
+        ----------
+        point_a : array-like, shape=[n_samples, dimension]
+            First point in hyperbolic space.
+        point_b : array-like, shape=[n_samples, dimension]
+            Second point in hyperbolic space.
+
+        Returns
+        -------
+        dist : array-like, shape=[n_samples, 1]
+            Geodesic distance between the two points.
+        """
+        point_a_norm = gs.clip(gs.sum(point_a ** 2, -1), 0., 1 - EPSILON)
+        point_b_norm = gs.clip(gs.sum(point_b ** 2, -1), 0., 1 - EPSILON)
+
+        diff_norm = gs.sum((point_a - point_b) ** 2, -1)
+        norm_function = 1 + 2 * \
+            diff_norm / ((1 - point_a_norm) * (1 - point_b_norm))
+
+        dist = gs.log(norm_function + gs.sqrt(norm_function ** 2 - 1))
+        dist = gs.to_ndarray(dist, to_ndim=1)
+        dist = gs.to_ndarray(dist, to_ndim=2, axis=1)
+        dist *= self.scale
+        return dist
+
+    def retraction(self, tangent_vec, base_point):
```

Comment:
`self` is not used? Make this a static method then?

Reply:
Sorry, I maybe forget in severall function to first check if base_point belong to the manifold.